### PR TITLE
Summer 26 Housekeeping: tests, docs, CI, and bug fixes

### LINKED ISSUE #331: Logging handling is not multiprocess safe
Current PyOPIA implementation of logging to a single file is not guaranteed to work with multiprocessing, which may lead to log file corruption. If this problem shows up in practice, it can be fixed by implementing a log queue.

See: https://docs.python.org/3/howto/logging-cookbook.html#logging-to-a-single-file-from-multiple-processes

### LINKED ISSUE #350: Using incorrectly formatted auxilliary data file makes pyopia skip image with non-useful log message
**Describe the bug**
Through copying a project folder from an older version, I ended up with an incorrectly formatted auxiliary data file. This resulted in pyopia skipping at the output step. The rois were written, but the stats.nc file were not written. The log message was

```
2025-08-07 14:22:59 INFO MainProcess [pipeline.run] Running pipeline step: load
2025-08-07 14:22:59 INFO MainProcess [pipeline.run] Running pipeline step: correctbackground
2025-08-07 14:23:00 INFO MainProcess [pipeline.run] Running pipeline step: imageprep
2025-08-07 14:23:00 INFO MainProcess [pipeline.run] Running pipeline step: segmentation
2025-08-07 14:23:00 INFO MainProcess [process.segment] segment
2025-08-07 14:23:00 INFO MainProcess [process.segment] clean
2025-08-07 14:23:01 INFO MainProcess [pipeline.run] Running pipeline step: statextract
2025-08-07 14:23:01 INFO MainProcess [process.__call__] statextract
2025-08-07 14:23:01 INFO MainProcess [process.concentration_check] 1.0% saturation
2025-08-07 14:23:01 INFO MainProcess [process.statextract] measure
2025-08-07 14:23:01 INFO MainProcess [process.measure_particles]   367 particles found
2025-08-07 14:23:03 INFO MainProcess [process.extract_particles] EXTRACTING 366 IMAGES from 367
2025-08-07 14:23:03 INFO MainProcess [process.__call__] CalculateImageStats
2025-08-07 14:23:03 INFO MainProcess [pipeline.run] Running pipeline step: output
2025-08-07 14:23:03 WARNING MainProcess [cli.process_file_list] [red]An error occured in processing, skipping rest of pipeline and moving to next image.(chunk 0)
2025-08-07 14:23:03 ERROR MainProcess [cli.process_file_list] 'time'
```

The last line is from ```cli.process_file_list``` printing the error passed down from ```pipeline.run```, which in this case has something to do with the malformed auxiliary data file. 

It would be useful if the error message gave an indication of where the problem might be found.

In the interest of early discovery, it would also be useful if an error in the auxiliary data would cause the processing to stop. Since the auxiliary data file is the same for all images, it is not necessary to keep trying again at every image, never producing any output.

### LINKED ISSUE #399: Migrate docs build from requirements.txt/pip to uv
The main codebase already builds/tests with `uv` (see `build-and-test.yml`, `pypi.yml`), but the docs build still installs via `docs/requirements.txt` + pip (`.readthedocs.yaml`). Read the Docs now supports `uv` natively (April 2026 release).

Proposed: fold docs-only deps into `pyproject.toml`, delete `docs/requirements.txt`, and switch `.readthedocs.yaml` to `method: uv` / `command: sync`.

Raised by @emlynjdavies on #397.

### LINKED ISSUE #403: Streamline test suite with shared, session-scoped pytest fixtures
Raised by @emlynjdavies on #394. `pyopia/tests/conftest.py` (added in #394) uses `@pytest.fixture(scope='session')` to share expensive real setup (network download of the example model/image, a real end-to-end `process` run) across multiple tests instead of repeating it per test.

This differs from the existing pattern in `test_pipeline.py` and `test_classify.py`, where each test function independently opens its own `tempfile.TemporaryDirectory()` and downloads its own copy of example data via `pyopia.exampledata`, even when another test elsewhere needs the same thing. Concretely, `test_classify.py::test_match_to_database` and #394's `silcam_example_files` fixture each separately download the same classifier model.

Proposed: promote the shared download/setup fixtures to a top-level `conftest.py`, and replace manual `tempfile.TemporaryDirectory()` usage with pytest's built-in `tmp_path`/`tmp_path_factory` fixtures across `test_pipeline.py` and `test_classify.py`. Should reduce redundant downloads and pipeline runs, and cut test-suite runtime.

### LINKED ISSUE #405: get_j() feeds uninitialized memory into the Junge slope fit for zero-count bins
Found via a NumPy UserWarning surfaced in #404's CI (unrelated PR - just where it was noticed).

\`pyopia.statistics.get_j()\`:

\`\`\`python
p = np.polyfit(
    np.log(dias[ind]),
    np.log(number_distribution[ind], where=number_distribution[ind] > 0),
    1,
)
\`\`\`

\`where=\` without a matching \`out=\` tells NumPy to skip computing \`log()\` for non-positive values (guarding against \`log(0)\`/\`log(negative)\`), but the skipped positions in the result are left as **uninitialized memory** - whatever bytes happened to already be there - not zero, not NaN. That array is then fed directly into \`np.polyfit\` for the Junge slope fit.

\`number_distribution\` is a real particle-count distribution, so a zero-count bin in the 150-300 um fitting range is a completely normal situation (an empty size bin), not an edge case. Whenever that happens, the Junge slope calculation is currently corrupted by garbage at exactly those points, rather than excluding them as the \`where=\` clause seems to have intended.

\`get_j()\` is called from \`nc_vc_from_stats()\`, which returns \`junge_slope\` as one of its four outputs and ultimately populates the \`junge\` column in \`image_stats\` (see \`pyopia.process.CalculateImageStats\`) - so this affects real, user-facing output, not just an internal detail.

**Fix**: exclude non-positive \`number_distribution\` bins from the fit entirely, by folding \`number_distribution > 0\` into the \`ind\` mask used for both the x (\`log(dias)\`) and y (\`log(number_distribution)\`) arrays, rather than only masking the \`log()\` call itself.

No existing test coverage for \`get_j()\`.

### LINKED ISSUE #409: README audit: broken docs-build instructions, typos, stale dev targets
Audit of README.md found several issues. Most notably, the "Build docs locally" instructions are broken (verified by running them - the referenced folders don't exist); the correct command is a single `sphinx-build -b html docs/ docs/_build/html`. There's also a broken markdown link, a stray Windows-style path in a bash command, several typos, and inconsistent "PyOpia"/"PyOPIA" naming.

The "Development targets" section is stale and in places contradicts the current codebase (e.g. it says "no use of settings/config files," but the whole pipeline is now config-driven) - worth trimming or rewriting.

Also proposing a few small, concise additions for new-user clarity: a one-line "what is this" intro, a note on where `config.toml` comes from in the quick-start, and a slightly fuller explanation of why Docker is offered.

### LINKED ISSUE #414: Add automated license-compliance check for dependencies
Prompted by Emlyn's review comment on #404 flagging that DINOv2's pretrained weights were originally released under a non-commercial license (CC BY-NC 4.0) before later moving to Apache 2.0. Pinning the DINOv2 weights version (tracked separately, in review discussion on #404) addresses that specific case, but there's currently no automated check that a *declared* PyPI dependency doesn't carry a license that's incompatible with the intention of PyOPIA's own permissive BSD3 license.

**Proposal**: add `pip-licenses` as a dev dependency, plus a CI step that fails the build if any dependency's license matches GPL/AGPL/LGPL/Commons Clause/Non-Commercial/SSPL/Business Source License (using `--partial-match`, since these tools match license strings exactly by default, not as substrings).

**Findings from testing this locally**:
- `docutils` needs an explicit `--ignore-packages` exception - its PyPI metadata reports a compound license string that includes GPL, from a couple of bundled non-code files, even though the actual code is BSD.
- `cmocean` and `namex` (both real PyOPIA dependencies) don't declare any machine-readable license on PyPI at all. Manually confirmed MIT and Apache 2.0 respectively, directly from their source repos - this check can't verify that automatically since neither package declares it themselves.
- PyOPIA itself didn't declare its own license in packaging metadata either - fixing that (`license = "BSD-3-Clause"` in `pyproject.toml`) is included as part of this issue.

Note: this check only covers declared PyPI dependencies (`pyproject.toml`/`uv.lock`). It doesn't cover runtime-downloaded artifacts like the DINOv2 weights, which aren't a Python package at all - that risk needs a different mitigation (version pinning), discussed separately on #404.

### LINKED ISSUE #415: Pipeline.print_steps() references a non-existent attribute
\`Pipeline.print_steps()\` calls \`steps_to_string(self.steps)\`, but \`Pipeline.__init__\` only ever sets \`self.settings\` and \`self.stepnames\` - there is no \`self.steps\` attribute, so this raises \`AttributeError\` if ever called. Currently unused anywhere in the codebase (no CLI/test/notebook calls it), which is presumably why this hasn't been noticed.

Fix: use \`self.settings['steps']\` instead.

### LINKED ISSUE #420: Decide how to handle DINOv2 weight drift: scheduled check vs. pinned version
Split out from #404 per Emlyn's suggestion, to stop it blocking the rest of that PR's testing-infrastructure work.

Open question: `.github/workflows/scheduled-training-check.yml` (added in #404) currently re-downloads and re-trains against whatever DINOv2 weights `timm` resolves to on a weekly schedule - Emlyn flagged a licensing risk in this, since DINOv2 itself started under a non-commercial license before later moving to Apache 2.0, and an unpinned weekly pull wouldn't catch a similar future change.

A ready fix is prepared on branch `404-pin-dinov2-weights-version`: pins `timm` to a specific version so the weights stay fixed until a maintainer deliberately bumps it (verified working). That approach would also remove the need for the scheduled workflow itself, per Emlyn's original comment - though there's a separate, smaller benefit the schedule currently provides (catching the notebook's own code going stale, independent of weights, as happened previously with `pipeline_step_by_step.ipynb`).

@emlynjdavies @nepstad - once there's agreement on direction, this is ready to implement quickly.

### LINKED ISSUE #428: Develop a Release philosophy, guidance and schedule (ish) for PyOPIA
See #425.

This could be written up as a new, brief section in the README file.

@emlynjdavies @animmosmith

### PR DESCRIPTION
**Pull request by animmosmith**
Combined housekeeping PR for the **Summer 26 Housekeeping** milestone, following up on our meeting with Emlyn and Raymond: a single reviewable PR covering tests/docs/bugs, ahead of a new release.

Each bullet below links to the specific commit that makes that change, so you can review them individually rather than as one combined diff.

## Bug fixes
- Closes #331, Closes #350: multi-chunk logging is now queue-based and auxiliary data errors are surfaced clearly, instead of being silently swallowed. ([`52b92da`](https://github.com/SINTEF/pyopia/pull/430/commits/52b92da55a58d392fa44aae1fd552cc50d0acd31))
- Closes #405: `get_j()` no longer feeds uninitialized memory into the Junge slope fit for zero-count bins. ([`ab815d2`](https://github.com/SINTEF/pyopia/pull/430/commits/ab815d285fa0ac3d830afffdf748e56e1261757c))
- Closes #415: `Pipeline.print_steps()` no longer references a non-existent attribute. ([`078d59c`](https://github.com/SINTEF/pyopia/pull/430/commits/078d59c365d4ef91eef88f3d6637c9a8db698bb6))
- Fixes a NumPy 2.5 shape-assignment deprecation warning in the holo reconstruction kernel (found via CI). ([`6ef82c1`](https://github.com/SINTEF/pyopia/pull/430/commits/6ef82c142b4d67862c5df2c7224bc7eb9c50aa0c))
- Ensures the queue log listener is always stopped on early validation errors ([`cb92b6f`](https://github.com/SINTEF/pyopia/pull/430/commits/cb92b6f6c985195238c4a0a1b63e2e3d3f3095c7)), and that the `multiprocessing.Queue` is fully closed after stopping the listener - prevents a Windows-specific CI hang ([`1aba960`](https://github.com/SINTEF/pyopia/pull/430/commits/1aba9600e373dc6957e82e3b0132267284051f26)).

## Tests
- Closes #403, and folds in #237: shared, session-scoped pytest fixtures, a `slow` marker, and per-notebook parametrized tests (including docs/notebooks coverage). ([`8206754`](https://github.com/SINTEF/pyopia/pull/430/commits/82067544bd03cc1358af1ea7131addbfc9bac706))
- Auto-retries notebook tests on failure ([`f341b78`](https://github.com/SINTEF/pyopia/pull/430/commits/f341b78f1028de024283977017d3fbdbc62c83ac)) and raises their timeout, based on real flakiness/timing evidence from CI ([`eeeab7d`](https://github.com/SINTEF/pyopia/pull/430/commits/eeeab7d8c3f346d6ecdf77643cb210e312230063)).

## CI
- Closes #420: bounds the DINOv2 training notebook's validation set (`VAL_PER_CLASS`, randomly sampled, instead of every leftover image) - cut the full 30-epoch default run from 40+ minutes (dominated by a one-time, unbounded ~7,400-image validation pass) to a few minutes on GitHub's runners. Moves the notebook from `pytest.mark.training` (never ran in CI) into the routine per-PR notebook suite, so a real training run now has to pass before merging. ([`a75c651`](https://github.com/SINTEF/pyopia/pull/430/commits/a75c651ff689884c15267aed74970938880552e4))
- Moves the notebook's `timm`/`scikit-learn` dependencies into the `classification-torch` extra, alongside `torch`, so `uv sync` resolves everything together respecting the project's own constraints (e.g. `click<8.2.0`) instead of an ad-hoc `!uv pip install` mutating the shared CI job's environment - confirmed via a real CI run that this was breaking `cli.ipynb`'s test in the same job. ([`444c9ae`](https://github.com/SINTEF/pyopia/pull/430/commits/444c9ae07be24e8b6556ec393ea198798e6d0a9a))
- Removes the now-unused `training` pytest marker entirely (nothing carries it once the above landed) rather than leaving dead config around; simplifies CI's invocation from `pytest -v -m "not training"` to `pytest -v`. ([`bbb4a25`](https://github.com/SINTEF/pyopia/pull/430/commits/bbb4a25f810b67752a23f8c32b0ac5607971bbf9))
- Pins `timm==1.0.28` exactly (was `timm>=1.0.0`) in both `pyproject.toml` and the notebook's standalone-install fallback: this closes the licensing half of #420 that the validation-cap fix above didn't address - DINOv2's pretrained weights started under a non-commercial license before later moving to Apache 2.0, and an unpinned `timm` (no `uv.lock` is committed) could silently resolve a future release onto different weights/licensing terms on every CI run. ([`658bd20`](https://github.com/SINTEF/pyopia/pull/430/commits/658bd2097054b70d47af6859280185d7fe2c0580))

## Docs
- Closes #399: migrates the docs build from requirements.txt/pip to uv. ([`42a9c13`](https://github.com/SINTEF/pyopia/pull/430/commits/42a9c13d333aa9ef09dbc781d56eaa5cf5f8ffe5))
- Closes #409: fixes a broken docs-build command, a broken link, typos, and stale content in the README. ([`68bb706`](https://github.com/SINTEF/pyopia/pull/430/commits/68bb706b9c5272a31542699acfae41348c23e753))
- Updates the local docs-build instructions to match the uv migration. ([`2674332`](https://github.com/SINTEF/pyopia/pull/430/commits/267433266795dd92a819d3b27bd3dd864a3eb766))

## Dependencies
- Closes #414: adds an automated license-compliance check for dependencies. ([`9772f87`](https://github.com/SINTEF/pyopia/pull/430/commits/9772f87b26ad7fbdd7338efc5f5c2ae36407ec73))

## Process
- Closes #428: adds a **Releases** section to the README documenting the release philosophy discussed in #425 - cut releases when work is finished and documented, rather than on a fixed schedule. ([`97bce17`](https://github.com/SINTEF/pyopia/pull/430/commits/97bce179eb31d47f34f89f4966f76dce1ac954de))
- Closes #429: adds a **Using AI tools** section to the README - human-facing policy: AI tools are welcome as an aid, but whoever submits a PR is responsible for the code in it ([`97bce17`](https://github.com/SINTEF/pyopia/pull/430/commits/97bce179eb31d47f34f89f4966f76dce1ac954de)) - and an `AGENTS.md` - agent-facing operational guidance: setup, testing, linting, conventions ([`bbb4a25`](https://github.com/SINTEF/pyopia/pull/430/commits/bbb4a25f810b67752a23f8c32b0ac5607971bbf9)).

## Test plan
- [x] `flake8 pyopia` clean
- [x] `uv run pytest -m "not slow"` - 36 passed
- [x] CI green on this PR (Windows/Ubuntu/macOS/License_check)

### CODE REVIEW

**Review by emlynjdavies — DISMISSED** [review]

**Comment by animmosmith** [comment]
Per Raymond's feedback at today's meeting: the DINOv2 training notebook is back out of routine CI.

- Reverted the exact `timm` pin ([`cc060b7`](https://github.com/SINTEF/pyopia/pull/430/commits/cc060b7bcbad626ca5ee0a2ed2ca1a7694699c31)) and the `torch`/`timm`/`scikit-learn` dependency move ([`d1a40ab`](https://github.com/SINTEF/pyopia/pull/430/commits/d1a40abb66bd890f4f57ac21ebd99bb091c4f110)) - both only mattered while the notebook ran in CI.
- Dropped the notebook from `SLOW_NOTEBOOKS` entirely and reverted its validation-capping code ([`33796fd`](https://github.com/SINTEF/pyopia/pull/430/commits/33796fd3cecd4ed286ca2cb042dc8d21a1ffa9c7)) - it's simply not tested in CI now, rather than reintroducing a separate `training` marker/CI filter for one notebook.

**Review by emlynjdavies — APPROVED** [review]
